dev-tools/xtask/src/virtual_hardware.rs

```diff
@@ -0,0 +1,834 @@
+        .context("Invalid modinfo output")?
+        .lines()
+        .find_map(|line| {
+            let mut cols = line.trim().splitn(2, ' ');
+            let id = cols.next()?;
+            let desc = cols.next()?;
+            if !desc.contains("xde") {
+                return None;
+            }
+            return Some(id.to_string());
+        });
+
+    let Some(id) = id else {
+        println!("xde driver already unloaded");
+        return Ok(());
+    };
+    println!("unloading xde driver");
+
+    let mut cmd = Command::new(PFEXEC);
+    cmd.arg(MODUNLOAD);
+    cmd.arg("-i");
+    cmd.arg(id);
+    execute(cmd)?;
+    Ok(())
+}
+
+fn remove_softnpu_zone(npu_zone: &Utf8Path) -> Result<()> {
+    println!("ensuring softnpu zone destroyed");
+    let mut cmd = Command::new(PFEXEC);
+    cmd.arg(npu_zone);
+    cmd.args([
+        "destroy",
+        "sidecar",
+        "--omicron-zone",
+        "--ports",
+        "sc0_0,tfportrear0_0",
+        "--ports",
+        "sc0_1,tfportqsfp0_0",
+    ]);
+    execute(cmd)?;
+    Ok(())
+}
+
+fn remove_vnics() -> Result<()> {
+    delete_address("lo0/underlay")?;
+    delete_interface("sc0_1")?;
+    delete_vnic("sc0_1")?;
+
+    for i in 0..=1 {
+        let net = format!("net{i}");
+        let sc = format!("sc{i}_0");
+
+        delete_interface(&net)?;
+        delete_simnet(&net)?;
+        delete_simnet(&sc)?;
+    }
+
+    Ok(())
+}
+
+fn ensure_simulated_links(
+    physical_link: &str,
+    promiscuous_filter_off: bool,
+) -> Result<()> {
+    for i in 0..=1 {
+        let net = format!("net{i}");
+        let sc = format!("sc{i}_0");
+        if !simnet_exists(&net) {
+            create_simnet(&net)?;
+            create_simnet(&sc)?;
+            modify_simnet(&sc, &net)?;
+            set_linkprop(&sc, "mtu", "9000")?;
+        }
+        println!("Simnet {net}/{sc} exists");
+    }
+
+    let sc = "sc0_1".to_string();
+    if !vnic_exists(&sc) {
+        create_vnic(&sc, physical_link, PXA_MAC_DEFAULT)?;
+        if promiscuous_filter_off {
+            set_linkprop(&sc, "promisc-filtered", "off")?;
+        }
+    }
+    println!("Vnic {sc} exists");
+    Ok(())
+}
+
+fn ensure_softnpu_zone(npu_zone: &Utf8Path) -> Result<()> {
+    let zones = zoneadm_list()?;
+    if !zones.iter().any(|z| z == "sidecar_softnpu") {
+        if !npu_zone.exists() {
+            bail!("npu binary is not installed. Please re-run ./tools/install_prerequisites.sh");
+        }
+
+        let mut cmd = Command::new(PFEXEC);
+        cmd.arg(npu_zone);
+        cmd.args([
+            "create",
+            "sidecar",
+            "--omicron-zone",
+            "--ports",
+            "sc0_0,tfportrear0_0",
+            "--ports",
+            "sc0_1,tfportqsfp0_0",
+            "--sidecar-lite-commit",
+            SIDECAR_LITE_COMMIT,
+            "--softnpu-commit",
+            SOFTNPU_COMMIT,
+        ]);
+        execute(cmd)?;
+    }
+
+    Ok(())
+}
+
+fn initialize_softnpu_zone(
```

Comment:
Compare with:

https://github.com/oxidecomputer/omicron/blob/d7d70d8366ee106d49cc0c7bfae2c6aa499ecf3b/tools/scrimlet/softnpu-init.sh#L59-L67

(I did make "pxa_start" and "pxa_end" optional, but if one is supplied, the other must be supplied)